lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -168,36 +237,133 @@ data BalanceInsufficientError = BalanceInsufficientError
+        pure $ Left $ InsufficientMinCoinValues $
+            NE.fromList insufficientMinCoinValues
+
+    | otherwise = do
+        state <- runSelection extraCoinSource utxoAvailable balanceRequired
+        let predictedChange = predictChange (selected state)
+        makeChangeRepeatedly predictedChange state <&> \case
+            Nothing -> -- Running out of ada-only UTxO to pay for fee / deposits
+                Left UnableToConstructChange
+            Just result ->
+                Right result
   where
+    SelectionCriteria{outputsToCover, utxoAvailable, extraCoinSource} = criteria
+
+    mkInputsSelected :: UTxOIndex -> NonEmpty (TxIn, TxOut)
+    mkInputsSelected =
+        fromMaybe invariantSelectAnyInputs . NE.nonEmpty . UTxOIndex.toList
+
     balanceAvailable :: TokenBundle
-    balanceAvailable = view #balance utxoAvailable
+    balanceAvailable = availableBalance utxoAvailable extraCoinSource
 
     balanceRequired :: TokenBundle
     balanceRequired = F.foldMap (view #tokens) outputsToCover
 
-    mkResult :: SelectionState -> SelectionResult
-    mkResult SelectionState {selected, leftover} =
-        case NE.nonEmpty (UTxOIndex.toList selected) of
-            Nothing ->
-                unableToSelectAnyInputsError
-            Just inputsSelected ->
+    insufficientMinCoinValues :: [InsufficientMinCoinValueError]
+    insufficientMinCoinValues =
+        mapMaybe mkInsufficientMinCoinValueError outputsToCover
+      where
+        mkInsufficientMinCoinValueError
+            :: TxOut
+            -> Maybe InsufficientMinCoinValueError
+        mkInsufficientMinCoinValueError o
+            | view (#tokens . #coin) o >= expectedMinCoinValue =
+                Nothing
+            | otherwise =
+                Just $ InsufficientMinCoinValueError
+                    { expectedMinCoinValue, insufficientlyCoveredOutput = o }
+          where
+            expectedMinCoinValue = minCoinValueFor (view (#tokens . #tokens) o)
+
+    -- There's a chicken-and-egg situation when it comes to calculating
+    -- transaction fees. On the one hand, we need to know the shape of the final
+    -- transaction to calculate its cost. But in order to construct the
+    -- transaction, we need to know what its cost is.
+    --
+    -- So, in order to not duplicate the logic from 'makeChange', we first
+    -- calculate a pre-selection considering the case where we have no fees to
+    -- pay, and no minimum value. This is guaranteed to succeed and will yield
+    -- the final shape of the change outputs (modulo amounts).
+    --
+    -- From there, we can successively re-iterate and construct a final
+    -- unbalanced selection that is leaving enough money out for fees. For this
+    -- method to work (efficiently), the function that estimates the cost of a
+    -- selection MUST NOT depend on change token quantity. That is, increasing
+    -- the token quantities of a change output must not make it more expensive.
+    predictChange
```

Comment:
I really like the motivation, but I think this comment is still missing an overview of what the `predictChange` function does.

If I call `predictChange` with a `UTxOIndex`, what can I expect to get back?

I wonder, would it be possible to add:

- a short summary stating what the function does, along the lines of "This function creates ...",
- a statement of the post-conditions and properties?